src/databricks/labs/ucx/source_code/pyspark.py

```diff
@@ -1,10 +1,155 @@
+        return (
+            isinstance(node, ast.Call)
+            and isinstance(node.func, ast.Attribute)
+            and self.min_args <= len(node.args) <= self.max_args
+            and isinstance(node.args[self.table_arg_index], ast.Constant)
+        )
+
+    @abstractmethod
+    def lint(self, from_table: FromTable, node: ast.Call) -> Iterator[Advice]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def apply(self, from_table: FromTable, node: ast.Call) -> None:
+        raise NotImplementedError()
+
+
+@dataclass
+class QueryMatcher(Matcher):
+
+    def lint(self, from_table: FromTable, node: ast.Call) -> Iterator[Advice]:
+        table_arg = node.args[self.table_arg_index]
+        assert isinstance(table_arg, ast.Constant)
+        for advice in from_table.lint(table_arg.value):
+            yield advice.replace(
+                start_line=node.lineno,
+                start_col=node.col_offset,
+                end_line=node.end_lineno,
+                end_col=node.end_col_offset,
+            )
+
+    def apply(self, from_table: FromTable, node: ast.Call) -> None:
+        table_arg = node.args[self.table_arg_index]
+        assert isinstance(table_arg, ast.Constant)
+        new_query = from_table.apply(table_arg.value)
+        table_arg.value = new_query
+
+
+@dataclass
+class TableNameMatcher(Matcher):
+
+    def lint(self, from_table: FromTable, node: ast.Call) -> Iterator[Advice]:
+        table_arg = node.args[self.table_arg_index]
+        assert isinstance(table_arg, ast.Constant)
+        dst = self._find_dest(from_table, table_arg.value)
+        if dst is not None:
+            yield Deprecation(
+                code='table-migrate',
+                message=f"Table {table_arg.value} is migrated to {dst.destination()} in Unity Catalog",
+                # SQLGlot does not propagate tokens yet. See https://github.com/tobymao/sqlglot/issues/3159
+                start_line=node.lineno,
+                start_col=node.col_offset,
+                end_line=node.end_lineno or 0,
+                end_col=node.end_col_offset or 0,
+            )
+
+    def apply(self, from_table: FromTable, node: ast.Call) -> None:
+        table_arg = node.args[self.table_arg_index]
+        assert isinstance(table_arg, ast.Constant)
+        dst = self._find_dest(from_table, table_arg.value)
+        if dst is not None:
+            table_arg.value = dst.destination()
+
+    @staticmethod
+    def _find_dest(from_table: FromTable, value: str):
+        parts = value.split(".")
+        return None if len(parts) != 2 else from_table.index().get(parts[0], parts[1])
+
+
+# see https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.SparkSession.html
+SPARK_SESSION_MATCHERS = [QueryMatcher("sql", 1, 1000, 0), TableNameMatcher("table", 1, 1, 0)]
+
+# see https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.Catalog.html
+SPARK_CATALOG_MATCHERS = [
```

Comment:
please use class fields instead of global variables and pass in `SparkSql` (or `Index`) as the first argument to matchers. This way we don't have to leak the `index` property (as added in `queries.py`) and have more homogenous common interfaces and freedom to inject more via constructor.

Existing example - a factory of PermissionsManager

Reply:
done